File: ts/src/day05.ts
import { DayBase } from './day-runner.js';

 // https://adventofcode.com/2021/day/5

export class Day05 extends DayBase {
    constructor() {
        super(5);
    }

    public async run(): Promise<void> {
        // input format: num,num -> num,num
        const input = await this.getInputLines();
        // const input = ['0,9 -> 5,9', '8,0 -> 0,8', '9,4 -> 3,4', '2,2 -> 2,1', '7,0 -> 7,4', '6,4 -> 2,0', '0,9 -> 2,9', '3,4 -> 1,4', '0,0 -> 8,8', '5,5 -> 8,2', ];

        const lines = input.map(ln => {
            const sp = ln.split(' -> ');
            const from = sp[0].trim().split(',');
            const to = sp[1].trim().split(',');
            return { from: [parseInt(from[0], 10), parseInt(from[1], 10)], to: [parseInt(to[0], 10), parseInt(to[1], 10)]};
        });

        console.info(`Day ${this.day}`);
        console.info(`  Part 1:`, this.getIntersections(lines, false));
        console.info(`  Part 2:`, this.getIntersections(lines, true));
    }

    private getIntersections(lines: { from: number[], to: number[]}[], includeDiagonals: boolean): number {
        const mapKey = (x: number, y: number): string => `${x}_${y}`;

        let map = new Map<string, number>();
        lines.forEach(line => {
            const xmin = Math.min(line.from[0], line.to[0]);
            const xmax = Math.max(line.from[0], line.to[0]);
            const ymin = Math.min(line.from[1], line.to[1]);
            const ymax = Math.max(line.from[1], line.to[1]);
            if (xmin === xmax || ymin === ymax) {
                // Horiz and vert lines
                for (let x = xmin; x <= xmax; x++) {
                    for (let y = ymin; y <= ymax; y++) {
                        map.set(mapKey(x,y), (map.get(mapKey(x,y)) || 0) + 1);
                    }
                }
            } else if (includeDiagonals) {
                // Start from the point with the smaller x
                // Then decide if y should increase or decrease
                let pfrom = line.from[0] === xmin ? line.from : line.to;
                let pto = line.from[0] === xmin ? line.to : line.from;
                let dy = pfrom[1] < pto[1] ? 1 : -1;

                let range = xmax - xmin; // 45 degree diagonal => xrange == yrange
                for (let x = 0; x <= range; x++) {
                    let key = mapKey(pfrom[0] + x, pfrom[1] + (dy * x));
                    map.set(key, (map.get(key) || 0) + 1);
                }
            }
        });

        let result = 0;
        map.forEach(m => result += m > 1 ? 1 : 0); // sum points with > 1 line passing through
        return result;
    }
}

File: py/day05.py
from collections import Counter

# https://adventofcode.com/2021/day/5

def intersectionCount(lineSegments, includeDiagonals):
    counts = Counter()
    for seg in lineSegments:
        xmin, xmax = min(seg[0][0], seg[1][0]), max(seg[0][0], seg[1][0])
        ymin, ymax = min(seg[0][1], seg[1][1]), max(seg[0][1], seg[1][1])
        if xmin == xmax or ymin == ymax: # straight lines
            for x in range(xmin, xmax + 1):
                for y in range(ymin, ymax + 1):
                    counts[(x,y)] += 1
        elif includeDiagonals:
            # which point (to or from) has the smaller x?
            pfrom = seg[0] if seg[0][0] == xmin else seg[1]
            pto = seg[1] if seg[0][0] == xmin else seg[0]
            dy = 1 if pfrom[1] < pto[1] else -1 # line goes up l->r or down?
            for x in range(xmax - xmin + 1): # for a 45° diagonal line x and y increment by the same amount
                counts[(pfrom[0]+x, pfrom[1]+(dy*x))] += 1

    return sum(1 for v in counts.values() if v > 1)

with open("../input/day05-input", "r") as f:
    input = [x.split(' -> ') for x in [line.strip() for line in f.readlines()]]
segments = []
for seg in input:
    x1,y1 = [int(s) for s in seg[0].split(',')]
    x2,y2 = [int(s) for s in seg[1].split(',')]
    segments.append([(x1,y1), (x2,y2)])

print("Day 5")
print("  Part 1:", intersectionCount(segments, False))
print("  Part 2:", intersectionCount(segments, True))
